File: from-0901-to-1000/0948-bag-of-tokens/solution.go
func bagOfTokensScore(tokens []int, power int) int {
    sort.Ints(tokens)
    score, maxScore := 0, 0

    for len(tokens) > 0 {
        if power >= tokens[0] {
            power -= tokens[0]
            tokens = tokens[1:]
            score += 1
            if score > maxScore {
                maxScore = score
            }
        } else if score > 0 {
            power += tokens[len(tokens) - 1]
            tokens = tokens[:len(tokens) - 1]
            score -= 1
        } else {
            break
        }
    }
    return maxScore
}

File: from-0901-to-1000/0948-bag-of-tokens/solution.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        q = deque(sorted(tokens))
        score, max_score = 0, 0
        
        while q:
            if power >= q[0]:
                power -= q.popleft()
                score += 1
                max_score = max(max_score, score)
            elif score > 0:
                power += q.pop()
                score -= 1
            else:
                break
                
        return max_score
        